src/tbb/global_control.cpp

```diff
@@ -139,13 +141,20 @@ class alignas(max_nfs_size) lifetime_control : public control_storage {
     }
 };
 
-static allowed_parallelism_control allowed_parallelism_ctl;
-static stack_size_control stack_size_ctl;
-static terminate_on_exception_control terminate_on_exception_ctl;
-static lifetime_control lifetime_ctl;
-static control_storage *controls[] = {&allowed_parallelism_ctl, &stack_size_ctl, &terminate_on_exception_ctl, &lifetime_ctl};
+static std::array<std::unique_ptr<control_storage>, 4> controls;
+static std::once_flag control_storage_flag;
```

Comment:
We have discussed the solution briefly within the team and it seems like it can result in more undefined behavior since `std::unique_ptr` and `std::once_flag` are not trivial structures and it can be unsafe to access them before initialization (in particular if `global_control_init()` would be called before constructing the `controls` and `control_storage_flag`.

We propose to have global static pointer to the storage containing all required global control variables and dynamically allocate it inside `governor::acquire_resources()` (which would be called during the first access to TBB) and deallocate inside `governor::release_resources ()`.
@pavelkumbrasev, feel free to add if I missed something.

What do you think about the proposed change?

Reply:
I am _very_ unfamiliar with TBB internal code. I suspected that this solution may introduce more initialization problems as you pointed out, but I did not know where TBB initialized and destroyed things... so `unique_ptr` was the "easy" answer. With your hint we can give it a try on the `governor::[acquire|release]_resources` functions.